internal/handlers/metrics.go

```diff
@@ -0,0 +1,126 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/canonical/metrics-k8s-proxy/internal/k8s"
+	"github.com/canonical/metrics-k8s-proxy/internal/util"
+)
+
+// HTTPClient defines the interface for the HTTP client.
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// RealHTTPClient is the production implementation of HTTPClient.
+type RealHTTPClient struct {
+	*http.Client
+}
+
+// MetricsHandler holds the HTTP client.
+type MetricsHandler struct {
+	client HTTPClient
+}
+
+// NewMetricsHandler creates a new MetricsHandler with the given HTTP client.
+func NewMetricsHandler(client HTTPClient) *MetricsHandler {
+	return &MetricsHandler{client: client}
+}
+
+// ScrapePodMetrics scrapes metrics from a given pod and returns the combined metrics with the "up" metric.
+// In case of errors, it logs them and returns the 'up=0' metric.
+func (h *MetricsHandler) ScrapePodMetrics(ctx context.Context, podIP string,
+	metricsEndpoint k8s.PodScrapeDetails) string {
+	hostPort := net.JoinHostPort(podIP, metricsEndpoint.Port)
+	url := fmt.Sprintf("http://%s%s", hostPort, metricsEndpoint.Path)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		// Log the error and return the 'up=0' metric
+		log.Printf("Error creating request for %s: %v", url, err)
+		return util.AppendUpMetric("", metricsEndpoint.PodName, metricsEndpoint.Namespace, 0)
+	}
+
+	resp, err := h.client.Do(req)
+	if err != nil {
+		// Log the error and return the 'up=0' metric
+		log.Printf("Error scraping %s: %v", url, err)
+		return util.AppendUpMetric("", metricsEndpoint.PodName, metricsEndpoint.Namespace, 0)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		// Log the error and return the 'up=0' metric for non-200 responses
+		log.Printf("Failed to scrape %s, status code: %d", url, resp.StatusCode)
+		return util.AppendUpMetric("", metricsEndpoint.PodName, metricsEndpoint.Namespace, 0)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		// Log the error and return the 'up=0' metric for body read errors
+		log.Printf("Error reading response from %s: %v", url, err)
+		return util.AppendUpMetric("", metricsEndpoint.PodName, metricsEndpoint.Namespace, 0)
+	}
+
+	// Append 'up=1' for successful scrape
+	labeledMetrics := util.AppendLabels(string(body), metricsEndpoint.PodName, metricsEndpoint.Namespace)
+	return util.AppendUpMetric(labeledMetrics, metricsEndpoint.PodName, metricsEndpoint.Namespace, 1)
+}
+
+// AggregateMetrics collects metrics from all pods concurrently and returns aggregated results.
+func (h *MetricsHandler) AggregateMetrics(ctx context.Context, pw *k8s.PodScrapeWatcher) []string {
+	var wg sync.WaitGroup
+	var respMu sync.Mutex
+	responses := []string{}
+
+	for podIP, metrics := range pw.PodMetricsEndpoints {
```

Comment:
I don't think this is safe wrt concurrency.  `pw.PodMetricsEndpoints` is shared and could be written to while we read it here.  Maybe there's some semantics about `range` I don't know, but otherwise I think we need a `pw.mu.Lock()` protecting this read (or maybe get a local copy of it before the loop?)